app/seeders/demo_data/work_package_seeder.rb

```diff
@@ -91,30 +91,35 @@ def create_children!(work_package, attributes)
 
         child.parent = work_package
         child.save!
+        memorize_work_package(child, child_attributes)
       end
     end
 
     def base_work_package_attributes(attributes)
       {
         project:,
         author: user,
-        assigned_to: find_principal(attributes['assignee']),
+        assigned_to: find_principal(attributes['assigned_to']),
         subject: attributes['subject'],
         description: attributes['description'],
         status: find_status(attributes),
         type: find_type(attributes),
         priority: find_priority(attributes) || IssuePriority.default,
-        parent: WorkPackage.find_by(subject: attributes['parent'])
+        parent: find_work_package(attributes['parent'])
       }
     end
 
-    def find_principal(name)
-      if name
-        group_assignee = Group.find_by(lastname: name)
-        return group_assignee unless group_assignee.nil?
-      end
+    def memorize_work_package(work_package, attributes)
+      project_data.store_reference(attributes['reference'], work_package)
+      attributes['work_package'] = work_package
```

Comment:
Yeah, it's ugly :) One would say pragmatic ;)

It's to create relations. Source looks like this:
```
          - start: 4
            t_subject: Invite attendees to conference
            reference: :invite_attendees_to_conference
            description: ''
            status: default_status_new
            type: default_type_task
            duration: 1
            relations:
              - to: :set_date_and_location_of_conference
                type: follows
            schedule_manually: false
```

The work packages are created, and then the relations. But when the relations are created, we do not have the instance of the work package anymore for the `from` part of the relation. That's why I stored it in the attributes hash. It is used [here](https://github.com/opf/openproject/pull/12466/files#diff-0142d833da37e098ead6067550b37a4883671edb1e4e3312f6920c60e8851930R165).

I could use the reference from the attributes to look up the `from` but not all work packages have a `reference`.

That's where I thought that I could automatically generate a `reference` from the name, but it would be a bit too "magical" and probably more complicated to handle name collisions. So I gave up on this idea. Maybe in a later iteration if necessary.

That's also where I thought

Reply:
Ok, I understand the rational now. 

Two ideas: 
* Instead of storing the work package in the attributes hash of that work package, extract the relations part into its own data structure/path within the `SeedData` object passed around. That object could either store the actual work package for the from part of the relation or just the reference (the symbol) of it. The code can then create all the work packages in one go which results in a collection of relations that still need to be created. Those can then be created in a separate step.
* Building on the idea above, when specifying the relations, they are no longer part of the work package attributes but have their own section just like work_packages. So the definition of the relation in the standard.yml would look like this:

```
            relations:
              - from: :invite_attendees_to_conference
                to: :set_date_and_location_of_conference
                type: follows
```

This would have the advantage of no longer having to look through the work packages to find the relations. On the other hand, one now would have to look up both work packages to see if the relation makes sense. I would still think that have a separate relations sections is both easy to implement now and simpler to maintain in the long run.

Reply:
Honestly, I'm not a big fan of having a separate relations section and having to create some references on work packages just to have a valid `from` to reference. Sure it's easier to implement, but from a seed file writer point of view, it makes it harder to author a relation between work packages or understand an existing file as one would have to jump back and forth between `work_packages` and `relations` sections.

To be honest, I even thought about reducing the relation to something shorter like this:

```yaml
        work_packages:
          - t_subject: Set date and location of conference
            reference: :set_date_and_location_of_conference
          - t_subject: Invite attendees to conference
            relations:
              - follows: :set_date_and_location_of_conference
```

Your proposition is

```yaml
        work_packages:
          - t_subject: Set date and location of conference
            reference: :set_date_and_location_of_conference
          - t_subject: Invite attendees to conference
            reference: :invite_attendees_to_conference
        relations:
          - from: :invite_attendees_to_conference
            to: ::set_date_and_location_of_conference
            type: follows
```

That works and is easier to implement, but I think it's harder to read and write and it feels redundant.

That discussion got me another idea. I'll commit it and you'll tell me what you think about it.

Reply:
I'll do it in another PR